### PR TITLE
Support LRU and WeakRef cache types

### PR DESCRIPTION
**Pull request by cevr**
This PR outlines a possible implementation of a generic cache, where the user can define their eviction policy, and we reconcile.

@bvaughn I need some help with tests failing for useCacheMutation 😓 

What are your thoughts?

### CODE REVIEW

**Comment by cevr** [comment]
@bvaughn Understood! Currently the default is a no-GC cache (simple new Map), the LRU cache can be provided by the user.  We can provide a WeakRef based cache as a utility the user can import and provide in config option `getCache`. What do you think?

**Comment by cevr** [comment]
@bvaughn You actually already had a `WeakRefMap` implementation made. Were you planning for this use case? Incredible

**Comment by bvaughn** [comment]
> @bvaughn You actually already had a WeakRefMap implementation made. Were you planning for this use case?

Not exactly. I had originally written the cache to use either `Map` or `WeakRefMap` based on the configuration. I ended up changing this (in 76ad0d1) when I added support for the mutation hook though, because it wasn't possible for a `FinalizationRegistry` callback to access the cache maps managed by React (`unstable_getCacheForType`) – so it wasn't possible to evict a GCed value that way.

**Review by bvaughn — COMMENTED** [review]
This direction seems promising! 😄  Left some questions.

I made some small structural changes to deferred and record. Sorry for causing a rebase headache. Hopefully shouldn't impact many lines.

**Comment by bvaughn** [comment]
> @bvaughn I need some help with tests failing for useCacheMutation 😓 

Oops. Just seeing this. Still need help?

File: packages/suspense/src/types.ts [hunk 1]
```diff
@@ -140,3 +139,15 @@ export interface StreamingCache<
   prefetch(...params: Params): void;
   stream(...params: Params): StreamingValue<Value, AdditionalData>;
 }
+
+export interface Cacher<Key, Value> {
```
[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense/src/types.ts:143]: Much better

File: packages/suspense/src/utils/WeakRefMap.ts [hunk 2]
```diff
@@ -47,18 +52,34 @@ export class WeakRefMap<Key, Value extends Object> {
...
+    this.map.clear();
+  }
+
+  get size(): number {
+    return this.map.size;
+  }
+
+  forEach(callback: (value: Value, key: Key, map: this) => void): void {
+    this.map.forEach((weakRef, key) => {
+      const value = weakRef.deref();
```
[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense/src/utils/WeakRefMap.ts:81]: Agreed, opting out of foreach seems like the more practical approach.

**Comment by cevr** [comment]
@bvaughn Seems like your suggestions made it so the tests passed again. Though now 1 test is failing, `line: 328` of `createCache.test.ts` because of a maximum call stack exceeded 🤔 . Having trouble intuiting the cause, any thoughts?

File: packages/suspense/src/utils/WeakRefMap.ts [hunk 3]
```diff
@@ -47,18 +52,34 @@ export class WeakRefMap<Key, Value extends Object> {
...
+    return this;
+  }
+
+  clear(): void {
+    this.map.clear();
+  }
+
+  get size(): number {
+    return this.map.size;
+  }
```
[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense/src/utils/WeakRefMap.ts]: From what I can tell, we use it internally

File: packages/suspense/src/cache/createCache.ts [hunk 4]
```diff
@@ -228,14 +238,6 @@ export function createCache<Params extends Array<any>, Value>(
 
     if (!record) {
       return STATUS_NOT_FOUND;
-    } else if (isResolvedRecord(record)) {
-      try {
-        readRecordValueThrowsIfGC(record);
-      } catch (error) {
-        evict(...params);
-
-        return STATUS_NOT_FOUND;
```
[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense/src/cache/createCache.ts]: You're right, thanks for catching that

File: packages/suspense/src/utils/WeakRefMap.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/suspense/src/utils/WeakRefMap.ts]: Fair, but we only use it for the return value of `evictAll` – and I think we could just change the signature of that method to not return a boolean and side step this issue.

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense/src/utils/WeakRefMap.ts]: Sounds good to me

File: packages/suspense/src/cache/createCache.test.ts [hunk 5]
```diff
@@ -251,6 +252,14 @@ describe("createCache", () => {
 
       expect(load).toHaveBeenCalledTimes(1);
     });
+
+    it("should error if the load function syncronously resolves to undefined", async () => {
```
Comment (bvaughn) [inline packages/suspense/src/cache/createCache.test.ts:256]: Ah, good test condition 👍🏼

File: packages/suspense/src/cache/createCache.ts [hunk 6]
```diff
@@ -357,18 +388,14 @@ export function createCache<Params extends Array<any>, Value>(
...
-        return retainedValue;
-      }
-    } else {
-      return value!;
+  function readRecordValueThrowsIfSilentlyEvicted(
+    record: ResolvedRecord<Value>
+  ): Value {
+    const { value } = record.data;
+    if (value == null) {
+      throw Error("Value was silently evicted");
```
Comment (bvaughn) [inline packages/suspense/src/cache/createCache.ts:396]: This looks unsafe to me.

A cache can reasonably resolve a value to `null` or `undefined` without it meaning something was garbage collected.

I think the `CacheMap` implementation needs to be the thing that throws on read, for this to be safe. Here's my reasoning:

* Normal `Map` implementation can bypass the overhead of this check entirely.
* LRU implementation can check its internal `has()` method for this.
* `WeakRef` implementation _can_ through if `value.deref()` is null, only because no non-object values should be stored in it to begin with.

File: packages/suspense/src/utils/WeakRefMap.ts [hunk 7]
```diff
@@ -1,14 +1,19 @@
...
+// WeakRefs only work with objects, yet the user may be specifying an arbitrary type.
+// This doesn't actually matter since internally what is stored is Record<Value>
+// which is an object.
+// Currently we lie to the type system to make this work seamlessly.
+// TODO: Find a better way to do this.
+export class WeakRefMap<Key extends string, Value> {
   private finalizerCallback: FinalizerCallback<Key>;
   private finalizationRegistry: FinalizationRegistry<Key>;
-  private map: Map<Key, WeakRef<Value>>;
+  private map: Map<Key, WeakRef<any>>;
```
Comment (bvaughn) [inline packages/suspense/src/utils/WeakRefMap.ts:11]: If we're storing `Record` in the map, we should type this as:
```ts
private map: Map<Key, WeakRef<Record<Value>>>;
```

File: packages/suspense/src/cache/createCache.ts (hunk 6 above)
Comment (cevr) [reply packages/suspense/src/cache/createCache.ts:396]: Ah I see what you mean. 
I think for most cases, people will be using `CacheMap` implementations that are hand rolled or from an npm package like `lru-cache`. 

With that in mind, what do you think about a heuristic similar to React where `null` is considered a valid resolved value while `undefined` is considered a miss? We can enforce that with types, and make this check stricter by checking only for `undefined`.

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.ts:396]: > With that in mind, what do you think about a heuristic similar to React where `null` is considered a valid resolved value while `undefined` is considered a miss?

I'm a bit uncomfortable with that as a built-in limitation of the library. I generally prefer using `null` over `undefined` for "non values" _but_ it's subjective and I know of at least one use case (at @replay– where this library is being used) where `undefined` _is a valid value_.

> We can enforce that with types, and make this check stricter by checking only for `undefined`.

Not everyone uses TypeScript, so I don't think we can/should rely on static type checking to enforce soundness either.

**Comment by bvaughn** [comment]
Let me know if you'd like a hand with any of the suggested changes. We're having a bit of an outage at my job at the moment and I have a little extra bandwidth.

**Comment by cevr** [comment]
@bvaughn I think we've been thinking about this from the wrong perspective. We don't actually need to check if the value is `nullish` at all! We store the whole record in the `CacheMap`, so if it's been silently evicted, the `getOrCreateRecord` method will create a brand new pending record for us whenever it is called. The onus is on the CacheMap to consistently notify of `evictions`. For LRU Caches this is easy, for the `WeakRefMap` this is where it might need more thought.

What do you think?

**Comment by cevr** [comment]
We could also use a data structure that mirrors the keys set within the `CacheMap`. If the data structure contains a key that the `CacheMap` does not, then it means it has been silently evicted and we can notify our subscribers of that change. The only issue is that silent evictions would only be noticed on reads unless we make some sort of polling based reconciler. 

I think this would be useful in dev so to notify the user that silent evictions are happening, while in prod that code is not present

**Comment by bvaughn** [comment]
> @bvaughn I think we've been thinking about this from the wrong perspective. We don't actually need to check if the value is `nullish` at all! We store the whole record in the `CacheMap`

Ah, right. The underlying value isn't what's being stored in the map. 👍🏼  I forgot that because it isn't how the original implementation worked.

> The onus is on the CacheMap to consistently notify of evictions.

Agreed. The complexity of this specific case is specific to `WeakRef` so the design should ideally rely on the `CacheMap` implementation to detect this case.

> We could also use a data structure that mirrors the keys set within the `CacheMap`. If the data structure contains a key that the `CacheMap` does not, then it means it has been silently evicted and we can notify our subscribers of that change. The only issue is that silent evictions would only be noticed on reads unless we make some sort of polling based reconciler. 
>
> I think this would be useful in dev so to notify the user that silent evictions are happening, while in prod that code is not present

A mirrored data structure sounds pretty complicated. I don't _think_ we need to go that far. Same for polling. `WeakRef` may GC at any time and it's not always detectable/observable by design– so I think that's okay. We don't need to try to build around that, so long as we handle the case (e.g. an evicted record on read). If someone chooses to use `WeakRef` to store values, I think they're basically accepting the limitations and mechanics of how that structure works.

**Comment by cevr** [comment]
@bvaughn Perfect! I don't have much more time today to work on this branch, so if you still have extra bandwidth, please feel free. 

Outside of the implementation, I haven't gotten around the docs, but I can do that sometime this weekend if you think there isn't more considerations to be had :D

File: packages/suspense/src/cache/createCache.ts [hunk 8]
```diff
@@ -148,6 +148,16 @@ export function createCache<Params extends Array<any>, Value>(
     return new Map();
```
Comment (cevr) [inline packages/suspense/src/cache/createCache.ts:148]: @bvaughn One thing I forgot on final pass is this function. Should this function also use the CacheMap? I don't really understand how `getCacheForType` works, so I'm a little hesitant to change anything.

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.ts:148]: Yes. If this type returned a normal `Map`, it would prevent GC. The `getCache` configuration option should be the thing passed to `getCacheForType`.

Comment (cevr) [reply packages/suspense/src/cache/createCache.ts:148]: hey @bvaughn, i've got a couple questions related to this:

1) Should these be kept in sync? If `recordMap` invalidates, should this be propagated to `backupRecordMap` and vice versa?
2) What is the relationship between these two caches? I've read the comments but still don't understand 😅

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.ts:148]: @cevr

I explain the way the two caches work in this Loom:
https://www.loom.com/share/55b645c9730a4958b8a2f91dfc19b72f

As for keeping the two "in sync" – the logic that was in `createCache` and `useCacheMutation` already should be preserved. It was intentional.

Comment (cevr) [reply packages/suspense/src/cache/createCache.ts:148]: Ahh I see! 

It seems like the map would only have records that are being mutated no? Is it fair to say it will only ever have a subset of what is in the `backupRecordMap`?

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.ts:148]: In almost every case, this would be a subset, yes.

Keep in mind there may be multiple copies of these (each a non-overlapping subset).

Comment (cevr) [reply packages/suspense/src/cache/createCache.ts:148]: It seems to me that the `backupRecord` map is less of a backup and more like the primary source of truth. 

If it uses a subset, then it might make sense to have it bypass the user defined `CacheMap`, since the eviction policies  would most likely be useless for the subsets. Otherwise, we risk conflicting evictions and maybe double notifications? Though it might be a non-issue since react would bail.

Tangentially I'm interested in the use case. Are these cache subsets primarily for notifying of pending mutations to components who did not initiate it? I've never personally have felt the need to know of pending mutations outside of the component, since if it fails it usually has no consequences in regards to the currently cached data. The flow I'm used to is something like mutate -> onSuccess -> (invalidate or update in-place) -> notify. The component would have easy access to tracking the status of the mutation since its in scope.

Comment (cevr) [reply packages/suspense/src/cache/createCache.ts:148]: Fair enough!

If the `backupMap` is the source of truth, should we just notify whenever that evicts, or notify for both?

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.ts:148]: That's a good question.

Do we care about evictions in the temporary subset caches? I'm not sure. We don't want to double-notify.

It probably won't matter much in practice. The only case where it would matter (I think?) would be a mutation that starts, but never finishes (or fails)? Then the record would only be in the "smaller" map?

Comment (cevr) [reply packages/suspense/src/cache/createCache.ts:148]: Yeah, agreed - it's probably unlikely to happen.

What do you think about renaming `backupRecordMap` to `recordMap` and the current `createRecordMap` to something like `createPendingMutationRecordMap`?

Comment (cevr) [reply packages/suspense/src/cache/createCache.ts]: This actually does cause an issue with the WeakRefMap. It seems like it creates a closure that react controls that prevents it from ever being fully collected 😩

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.ts]: > What do you think about renaming backupRecordMap to recordMap and the current createRecordMap to something like createPendingMutationRecordMap?

I think that sounds fine.

> This actually does cause an issue with the WeakRefMap. It seems like it creates a closure that react controls that prevents it from ever being fully collected 😩

Really 🤔  That's unexepcted.

File: packages/suspense/src/cache/createCache.test.ts [hunk 9]
```diff
@@ -647,7 +647,7 @@ describe("createCache", () => {
       gcCache.cache({ key: "test" }, "test");
       expect(gcCache.getValueIfCached("test")).toEqual({ key: "test" });
 
-      expect(weakRefArray.length).toBe(1);
+      expect(weakRefArray.length).toBe(2);
```
Comment (bvaughn) [inline packages/suspense/src/cache/createCache.test.ts:650]: I don't understand this commit. It looks kind of like it's "fixing" the test by changing the expectations to be values that wouldn't be expected? Am I misreading it?

Comment (bvaughn) [reply packages/suspense/src/cache/createCache.test.ts:650]: Oh! Haha, no need for apologies. I just added those two together and misinterpreted the situation 😁 

Okay, that makes sense, although the tests now look a bit odd at a glance:
```js
expect(weakRefArray.length).toBe(2); // Not obvious at a glance why 2
weakRefArray[0].collect(); // Not obvious why we only collect/evict 1; shouldn't we evict all?
```

I wonder if there's a way we can write them to be clearer? 😄

**Comment by bvaughn** [comment]
There were some conflicts in `main` so I merged.

**Comment by cevr** [comment]
@bvaughn What are your thoughts on next steps?

**Comment by bvaughn** [comment]
> What are your thoughts on next steps?

Need some time to dig in and think about it more. I'm kind of too tired after work today to do much, except tweak the docs a bit. I'll try to find time to do more soon.

File: packages/suspense-website/src/examples/createCache/cacheWithLRU.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+// REMOVE_BEFORE
+
+import LRUCache from "lru-cache";
+
+createCache<[userId: string], JSON>({
+  config: {
+    getCache: (onEvict) =>
+      new LRUCache<string, Record<JSON>>({
+        max: 100,
+        dispose: (value, key) => onEvict(key),
```
[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts:14]: > Secondly, I believe dispose is called _before_ it's actually removed from the cache. The docs say that it's removed in the next tick, so something like `setTimeout(() => onEvict(key), 1)` works as well

This doesn't seem necessary. There's no harm in evicting a frame earlier than the cache, is there?

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: The problem is that if we notify a frame earlier, then our subscribers will receive the unevicted status rather than the post eviction status

[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: Why?

Forgetting about the details of this LRU for a second– calling `evict` on a cache will _synchronously evict_ the record and the status will then be "not-found" for any subscribers. Why would the LRU's internal details change this?

(Sorry if I'm overlooking something...it's been a long day.)

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: I'm not sure why it's implemented this way but since dispose is called before it's actually evicted, calling the `onEvict` callback synchronously means our subscribers will notify of this particular lru cache _pre_ eviction, since the actual eviction will happen in the next frame. Meaning when it calls getStatus, it will appear to still be whatever it was pre eviction. 

Sorry about the long day, hope you get some rest

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: To further clarify, the way we synchronize with the cache provided by `getCache` is by providing an `onEvict` callback to the `getCache` function.

[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: > calling the `onEvict` callback synchronously means our subscribers will notify of this particular lru cache _pre_ eviction, since the actual eviction will happen in the next frame. Meaning when it calls getStatus, it will appear to still be whatever it was pre eviction. 

Ah. I think I understand. I have not dug into the implementation of how the `onEviction` callback worked. I assumed it was calling `evict` but it's actually calling a different method `onExternalCacheEviction`. Any reason for that? Were you worried that calling `evict` would be unnecessarily wasteful?

If we _did_ call `evict`– then timing sensitivity issues like this could be avoided.

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: Yeah, my reasoning was that these cache implementations would evict themselves when they needed to, so our task was just to keep up to date with the work done. Generally this isn't an issue when the onEvict is called after the eviction has occurred, but with this particular implementation it is a bit clunky

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: We could just call evict ourselves, and if the cache had already evicted when `onEvict` was called, then it's just a noop. Not sure if that would cause issues with this particular lru-cache implementation. Only reason I included it in the docs was because it's the most popular implementation i know of in the ecosystem

[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: > my reasoning was that these cache implementations would evict themselves when they needed to, so our task was just to keep up to date with the work done

Agreed, but I think little "foot guns" like forgetting a `setTimeout` are things that would be nice to avoid if possible :🤪

[the remark this answers is not in the document]
Comment (bvaughn) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: FWIW I updated our unit tests to use `lru-cache`. I'll add a test later this evening for the edge case we're talking about. Hopefully I'll have the time and mental energy to dig into this stuff more fully this weekend.

[the remark this answers is not in the document]
Comment (cevr) [reply packages/suspense-website/src/examples/createCache/cacheWithLRU.ts]: Awesome! I'm not sure if it's still issue now that they've released a new major. 

Have a great weekend :D happy sabbath 🙇

**Comment by bvaughn** [comment]
I think this looks solid. Thank you for your help!
